docs/spec/concepts.rst

```diff
@@ -3,97 +3,216 @@
+Python objects have an attribute ``foo``.
+
+A term typed as :ref:`Any`, on the other hand, should be assumed to have
+_some_ specific static type, but _which_ static type is not known. A static
+checker should not emit any errors that depend on assuming a particular static
+type; a static checker should instead assume that the runtime is responsible
+for checking the type of operations on this term, as in a dynamically-typed
+language.
+
+The gradual guarantee
+~~~~~~~~~~~~~~~~~~~~~
+
+:ref:`Any` allows gradually adding static types to a dynamically-typed program.
+For a fully dynamically-typed program, a static checker has the type :ref:`Any`
+for all terms, and should emit no errors. Adding more static type annotations
+to the program (making the program more statically typed) may result in static
+type errors, if the program is not correct or if the static type annotations
+aren't able to fully represent the runtime types. Removing static type
+annotations (making the program more dynamic) should not result in additional
+static type errors. This is often referred to as the **gradual guarantee**.
+
+In Python's type system, we don't take the gradual guarantee as a strict
+requirement, but it's a useful guideline.
+
+The subtype relation
+--------------------
+
+A static type ``B`` is a **subtype** of another static type ``A`` if and only
+if the set of values represented by ``B`` is a subset of the set of values
+represented by ``A``. Because the subset relation on sets is transitive and
+reflexive, the subtype relation is also transitive (if ``C`` is a subtype of
+``B`` and ``B`` is a subtype of ``A``, then ``C`` is a subtype of ``A``) and
+reflexive (``A`` is always a subtype of ``A``).
+
+The **supertype** relation is the inverse of subtype: ``A`` is a supertype of
+``B`` if and only if ``B`` is a subtype of ``A``; or equivalently, if and only
+if the set of values represented by ``A`` is a superset of the values
+represented by ``B``. The supertype relation is also transitive and reflexive.
+
+We also define an **equivalence** relation on static types: the types ``A`` and
+``B`` are equivalent (or "the same type") if and only if ``A`` is a subtype
+of ``B`` and ``B`` is a subtype of ``A``. This means that the set of values
+represented by ``A`` is both a superset and a subset of the values represented
+by ``B``, meaning ``A`` and ``B`` must represent the same set of values.
+
+We may describe a type ``B`` as "narrower" than a type ``A`` (or as a "strict
+subtype" of ``A``) if ``B`` is a subtype of ``A`` and ``B`` is not equivalent
+to ``A``.
+
+The consistency relation
+------------------------
+
+Since :ref:`Any` represents an unknown static type, it does not represent any
+known single set of values, and thus it is not in the domain of the subtype,
+supertype, or equivalence relations on static types described above. (We do
+find it convenient to say, however, that ``Any`` is both subtype and supertype
```

Comment:
I agree that this sentence is awkward. Many gradual type systems in the
literature do define Any as a subtype of itself. It makes simplification of
union types more straightforward to define. If we don't allow Any to be a
subtype of itself, then the union type `int | Any | Any` can't be simplified to
`int | Any` under the rule that we can simplify subtypes out of unions. (We
can't simplify unions based on assignability, or else we get e.g. `int | Any`
simplifying to `Any`, which we don't want.)

I think a clearer option here might be to say explicitly in our definition of
unions that `Any | Any` is equivalent to `Any`, because the union of two
unknown sets of objects is an unknown set of objects.

This makes more intuitive sense than saying `Any <: Any`, since we can't really
say that one unknown set of objects is a subset of another unknown set of
objects.

@superbobry reminded me in an earlier comment to consider the case of union
simplification, which led me to add this sentence. @superbobry, do you think it
is actually important that we define `Any <: Any`, or can we instead just
define `Any | Any == Any`?

Reply:
I cannot think of another case where it is important that `Any <: Any`, but maybe @kmillikin has an example.

Reply:
The latest version no longer says that `Any <: Any`, but instead says directly that `Any | Any` is `Any`.